### PR TITLE
Improve logging in ledger API server [DPP-231]

### PR DESCRIPTION
**Pull request by kamil-da**
**Motivation**
Troubleshooting is made unnecessarily hard by lack of adequate logging. Every client request should be logged at the info level as it enters and leaves the system.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by kamil-da** [comment]
@mziolekda Could you review changes and point out which information worth logging I haven't included in this draft?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:70]: Should you add something to the logging context here?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:57]: It looks like the submission identifier has already been put in the logging context in the first line of this method, is it not?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:62]: Is there some submission identifier that should be added to the logging context here?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:94]: Would it make more sense to rather have the `partyIdHint` here? Even though both are optional, the `displayName` has more relaxed uniqueness requirements and it's probably less useful to track information when debugging.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:96]: Ideally this kind of information should be part of the logging context. In this way, it can also be used both for normal and structured logging.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:297]:
```suggestion
    logger.info("Storing party entry")
```

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:468]:
```suggestion
    logger.info("Storing transaction events")
```

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:500]:
```suggestion
      logger.info("Storing transaction")
```

File: libs-scala/contextualized-logging/BUILD.bazel
Comment (stefanobaghino-da) [inline libs-scala/contextualized-logging/BUILD.bazel:41]: I don't think that this library should depend on Akka. Do you think you could move the helper to the only place where it's used?

**Review by stefanobaghino-da — COMMENTED** [review]
Thanks, it basically looks ok, I've added a few notes here and there.

File: libs-scala/contextualized-logging/BUILD.bazel
Comment (kamil-da) [reply libs-scala/contextualized-logging/BUILD.bazel:41]: It already depends on Akka (which is a bit strange, I agree). I've only added Akka to test suite dependencies. I could extract stream utils into a separate helper but I'm not sure if that's relevant for this PR. What do you think?

Comment (stefanobaghino-da) [reply libs-scala/contextualized-logging/BUILD.bazel:41]: Ha, you're right, thanks for making me notice. Feel free to split the Akka part off, but don't feel forced to. 🙂

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:96]: I agree - I just wasn't sure if that wouldn't cause context cluttered.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:70]: There's no identifier available in the request. Should we generate one in such case?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:57]: Good catch!

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:62]: The same as above - no identifier available. What to do in such case?

Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:62]: Nothing, really. I thought that was the case, but felt like commenting just to made sure. Thanks for confirming!

(same applies to other comments)

Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:62]: Got it, thanks!

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:96]: It's definitely not a problem, put in the context everything you think is valuable for post-hoc observability. Consider that this is printed on the console only when using the normal logging configuration, in production you would use a structured logging framework and the extra information would not be in the message.

Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:96]: Thanks for the explanation. I'll add the party hint to the context then 👍

File: libs-scala/contextualized-logging/BUILD.bazel
Comment (kamil-da) [reply libs-scala/contextualized-logging/BUILD.bazel:41]: I'll address that in a separate PR.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala]: Fixed.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala]: Fixed.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala]: Fixed.

**Review by stefanobaghino-da — APPROVED** [review]
LGTM (but maybe wait for someone from your team to approve as well, if you think that makes sense)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala
Comment (kamil-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala:84]: I would appreciate a comment on whether we want complete transactions and transaction trees logged or should we log only a selection of fields? @mziolekda

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala:57]: We need to dial down the verbosity of this a bit. All we really need is a command id and status. Running it through proto introduces  yet another naming flavour i.e. `command_id`

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala:84]: Thanks, it is very thoughtful of you.
Those logs end up being way to verbose when passed through proto. We only need transactionId, commandId, workflowId and offset under those very names.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: PartyEntry contains submissionId, please log it as well

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:459]: We can skip this log, the final one when completing is enough

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:470]: We can skip this log, the final one when completing is enough

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:501]: This message looks unnecessarily verbose in the logfile and every field is prefixed with `update`:
```
updateCommandId=PRPruneQueryEmptyRangeOk-alpha-39f3ccbfdb89-command-220, updateLedgerTime=2021-02-03T15:58:31.515Z, updateOffset=0000000000000a190000000000000000, updateSubmitter=[PRPruneQueryEmptyRangeOk-alpha-39f3ccbfdb89-party-0], transactionId=0A2463633665333436652D373965652D333034322D613862642D653164633433626661636463, updateTransactionId=0A2463633665333436652D373965652D333034322D613862642D653164633433626661636463})
```
Would it be possible to dial it down a bit?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:558]: Submitter info contains submissionId, please log it when provided

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:738]: PackageLedgerEntry contains submissionId, please log it when provided

**Review by mziolekda — CHANGES_REQUESTED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:501]: I would recommend keeping it as is. Consider that you only see this when using a console logger, in production you should use structured logging and this would not be duplicated in the message.

Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:501]: but why do we have the prefix, i.e. updateCommandId vs commandId everywhere else, and why do we have transactionId and updateTransactionId both with the same content?

Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:501]: Had a look at the code, the transaction identifier is injected in the context upstream, as soon as it becomes available, using `loggingContextFor` in the trait `com.daml.platform.indexer.ExecuteUpdate`. I believe there's a number of keys that is being inserted twice.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: Re: this and other comments.

Before acting on this you may want to check whether these are already injected in the logging context upstream.

See `ExecuteUpdate#loggingContextFor`

```scala
      case PartyAllocationRejected(submissionId, participantId, _, rejectionReason) =>
        Map(
          "updateSubmissionId" -> submissionId,
          "updateParticipantId" -> participantId,
          "updateRejectionReason" -> rejectionReason,
        )
      case PublicPackageUpload(_, sourceDescription, _, submissionId) =>
        Map(
          "updateSubmissionId" -> submissionId.getOrElse(""),
          "updateSourceDescription" -> sourceDescription.getOrElse(""),
        )
```

**Comment by stefanobaghino-da** [comment]
Judging from @mziolekda's most recent comments it appears that many of the keys that are being inserted in this PR were actually already there. It would probably be best to have a second look at this.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: The `ExecuteUpdate#loggingContextFor` method enriches logging context for data writes in `JdbcLedgerDao`. There's no need to enrich the logging context in the latter.
However, I think we should consider removing `update` prefix from the context keys to be consistent with other places where we log correlation ids (`commandId`, `transactionId`, `submissionId`). What do you think @stefanobaghino-da @mziolekda ?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:459]: Same comment as [above](https://github.com/digital-asset/daml/pull/8676#discussion_r571638038).

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:470]: Fixed.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:501]: Same comment as [above](https://github.com/digital-asset/daml/pull/8676#discussion_r571638038).

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:558]: Same comment as [above](https://github.com/digital-asset/daml/pull/8676#discussion_r571638038).

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:738]: Same comment as [above](https://github.com/digital-asset/daml/pull/8676#discussion_r571638038).

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala:84]: Fixed 👍

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala:57]: Fixed 👍

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: I am strongly for removing that prefix. It leaks a bit of architecture detail to the log (updateCommandId differs from commandId only by the place in the code that prints them). For an end-user or support personnel it can be confusing.

Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: Also, the chain of logging context must be broken somewhere, because the example you posted elsewhere did not contain a context:
```
INFO: Storing package entry
```

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:468]: It is still here :-)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala:66]:
```suggestion
      "statusCode" -> statusCode.mkString,
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala:63]: Does this log line bring enough in its current form? Should it be `info` ?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala:66]: Good catch!

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala:63]: That's a good question. There's no information in the request itself that is suitable for putting in the `LoggingContext`. We could create a correlation id at this point but according to @stefanobaghino-da 's comment we're fine with this simple logging. Please see [this thread](https://github.com/digital-asset/daml/pull/8676#discussion_r568441915).

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala:66]: Actually it's not possible with our current Scala compiler settings. Specifically, compilation will fail with:
```
error: [wartremover:Option2Iterable] Implicit conversion from Option to Iterable is disabled - use Option#toList instead
```
I could use `statusCode.toList.mkString` but it looks a bit strange to create a list only for `mkString`. What do you think?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: From what I see in the codebase, there's no broken chain of logging context but rather the same method for storing package entries is used in different places. In some of them the logging context is not used at all. However, for the cases pertinent to this PR, the logging context is preserved.

Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: > I am strongly for removing that prefix. It leaks a bit of architecture detail to the log (updateCommandId differs from commandId only by the place in the code that prints them). For an end-user or support personnel it can be confusing.

I've removed the prefix. The `LoggingContext`'s keys should now have uniform naming.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala]: Fixed fixed.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala
Comment (kamil-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala:350]: I've created this object to ensure consistent logging for different keys. This is very similar to `com.daml.platform.apiserver.services.logging`. I wonder if it would be worth merging both of these helper utils. What do you think @mziolekda ?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala:63]: those simple get* requests have very little meat on the bone. They are also very shallow i.e. don't reach to committer. It is ok to log little info here.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: Let's have a look at the example log-line from an actual run and close this thread if the context appears

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala:350]: I am ok with splitting it to a follow-up PR. This one is getting big as it stands.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:299]: An example log line for this case is:
```
2021-02-10T18:07:01.537Z INFO  c.d.p.s.d.JdbcLedgerDao@[kka.actor.default-dispatcher-6] -
                Storing package entry (context: {participantId=sandbox-participant})
```

**Review by mziolekda — APPROVED** [review]
It looks ok now, thanks

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala
Comment (oliverse-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:113]: @kamil-da , don't you feel logging every submission at info level might turn out to be a bit verbose?

Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:113]: @oliverse-da Thank you for the comment.
Actually I think it may be more reasonable to lower the level to debug.
This change was explicitly requested by @mziolekda , so his input might be valuable here as well.

**Pull request merged**
